codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketCloseFrameHandler.java

```diff
@@ -0,0 +1,95 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.websocketx;
+
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelOutboundHandlerAdapter;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.concurrent.ScheduledFuture;
+import io.netty.util.internal.ObjectUtil;
+
+import java.nio.channels.ClosedChannelException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Send {@link CloseWebSocketFrame} message on channel close, if close frame was not sent before.
+ */
+final class WebSocketCloseFrameHandler extends ChannelOutboundHandlerAdapter {
+    private final WebSocketCloseStatus closeStatus;
+    private final long forceCloseTimeoutMillis;
+    private ChannelPromise closeSent;
+
+    WebSocketCloseFrameHandler(WebSocketCloseStatus closeStatus, long forceCloseTimeoutMillis) {
+        this.closeStatus = ObjectUtil.checkNotNull(closeStatus, "closeStatus");
+        this.forceCloseTimeoutMillis = forceCloseTimeoutMillis;
+    }
+
+    @Override
+    public void close(final ChannelHandlerContext ctx, final ChannelPromise promise) throws Exception {
+        if (!ctx.channel().isActive()) {
+            ctx.close(promise);
+            return;
+        }
+        if (closeSent == null) {
+            write(ctx, new CloseWebSocketFrame(closeStatus), ctx.voidPromise());
+        }
+        flush(ctx);
+        applyCloseSentTimeout(ctx);
+        closeSent.addListener(new ChannelFutureListener() {
+            @Override
+            public void operationComplete(ChannelFuture future) {
+                ctx.close(promise);
+            }
+        });
+    }
+
+    @Override
+    public void write(final ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        if (closeSent != null) {
+            if (!promise.isVoid() && !promise.isDone()) {
+                promise.tryFailure(new ClosedChannelException());
+            }
+            return;
+        }
+        if (msg instanceof CloseWebSocketFrame) {
+            promise = promise.unvoid();
+            closeSent = promise;
+        }
+        super.write(ctx, msg, promise);
+    }
+
+    private void applyCloseSentTimeout(ChannelHandlerContext ctx) {
+        if (closeSent.isDone() || forceCloseTimeoutMillis < 0) {
+            return;
+        }
+
+        final ScheduledFuture<?> timeoutTask = ctx.executor().schedule(new Runnable() {
+            @Override
+            public void run() {
+                closeSent.tryFailure(new WebSocketHandshakeException("send close frame timed out"));
```

Comment:
nit: check `if (!closeSend.isDone()...` first to reduce the overhead in case there is a race

Reply:
fixed